cmake/embed_helpers.cmake

```diff
@@ -0,0 +1,88 @@
+# Detect the distribution bpftrace is being built on
+function(detect_os)
+  file(STRINGS "/etc/os-release" HOST_OS_INFO)
+
+  foreach(os_info IN LISTS HOST_OS_INFO)
+    if(os_info MATCHES "^ID=")
+      string(REPLACE "ID=" "" HOST_OS_ID ${os_info})
+      set(HOST_OS_ID ${HOST_OS_ID} PARENT_SCOPE)
+    elseif(os_info MATCHES "^ID_LIKE=")
+      string(REPLACE "ID_LIKE=" "" HOST_OS_ID_LIKE ${os_info})
+      set(HOST_OS_ID_LIKE ${HOST_OS_ID_LIKE} PARENT_SCOPE)
+    endif()
+  endforeach(os_info)
+endfunction(detect_os)
+
+function(prepare_patch_series patchSeries patchPath)
+  message("Writing patch series to ${patchPath}/series ...")
+  file(WRITE "${patchPath}/series" "")
+  foreach(patch_info IN ITEMS ${patchSeries})
+    file(APPEND "${patchPath}/series" "${patch_info}\n")
+  endforeach(patch_info)
+endfunction(prepare_patch_series patchSeries patchPath)
+
+function(fetch_patches patchName patchPath patchURL patchChecksum)
+  if(NOT EXISTS "${patchPath}/${patchName}")
+    message("Downloading ${DEBIAN_PATCH_URL}")
+    file(MAKE_DIRECTORY ${patchPath})
+    file(DOWNLOAD "${DEBIAN_PATCH_URL}" "${patchPath}/${patchName}"
+         EXPECTED_HASH SHA256=${patchChecksum})
+
+    # Can add to this if ladder to support additional patch formats, tar
+    # probably catches quit a lot...
+    if(patchName MATCHES .*tar.*)
+      execute_process(COMMAND tar -xpf ${patchPath}/${patchName} --strip-components=3 -C ${patchPath})
+    else()
+      message("Patch ${patchName} doesn't appear to a tar achive, assuming it is a plaintext patch")
+    endif()
+  endif()
+endfunction(fetch_patches patchName patchPatch patchURL patchChecksum)
+
+function(prepare_clang_patches)
+  message("Building embedded Clang against host LLVM, checking compatibiilty...")
+  detect_os()
+  message("HOST ID ${HOST_OS_ID}")
+
+  set(CLANG_PATCH_COMMAND "/bin/true")
+  if(HOST_OS_ID STREQUAL "debian" OR HOST_OS_ID STREQUAL "ubuntu" OR HOST_OS_ID_LIKE STREQUAL "debian")
+    message("Building on a debian-like system, will apply minimal debian patches to clang sources in order to build.")
+    set(PATCH_NAME "debian-patches.tar.gz")
+    set(PATCH_PATH "${CMAKE_CURRENT_BINARY_DIR}/debian-llvm/")
+    set(DEBIAN_PATCH_SERIES "")
+    list(APPEND DEBIAN_PATCH_SERIES "kfreebsd/clang_lib_Basic_Targets.diff -p2")
+
+    if(${LLVM_VERSION} VERSION_EQUAL "8" OR ${LLVM_VERSION} VERSION_GREATER "8" )
```

Comment:
I haven't tested patching LLVM 7 or 9, and didn't think there was any point doing this for old LLVM versions.